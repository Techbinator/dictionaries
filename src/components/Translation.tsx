import * as React from "react";
import {
  ITranslations,
  TremoveTranslation,
  TeditTranslation
} from "../store/Dictionaries.type";
import { validateTranslation } from "../helpers/validation";

export interface IDictionaryDataProps {
  dictionaryUUID: string;
  translationsUUID: string[];
  translationUUID: string;
  translations: ITranslations;
  removeTranslation: TremoveTranslation;
  editTranslation: TeditTranslation;
}
interface ITranslationState {
  domain: string;
  range: string;
  domainError: string;
  rangeError: string;
  severity: string;
}
export default class Translation extends React.Component<
  IDictionaryDataProps,
  ITranslationState
> {
  state = {
    domain: this.props.translations[this.props.translationUUID].domain,
    range: this.props.translations[this.props.translationUUID].range,
    domainError: "",
    rangeError: "",
    severity: ""
  };

  static getDerivedStateFromProps(
    props: IDictionaryDataProps,
    state: ITranslationState
  ) {
    const { translationsUUID, translations, translationUUID } = props;
    const { domain, range } = state;
    const validation = validateTranslation(
      domain,
      range,
      translationsUUID,
      translations,
      translationUUID
    );
    return validation;
  }

  editTranslationsHandler = () => {
    const { editTranslation, translationUUID } = this.props;
    const { domain, range } = this.state;

    editTranslation({
      uuid: translationUUID,
      domain: domain,
      range: range
    });
  };
  public render() {
    const { removeTranslation, dictionaryUUID, translationUUID } = this.props;
    const { domain, range, domainError, rangeError, severity } = this.state;
    return (
      <tr>
        <td style={{ width: "40%" }}>
          <input
            type="text"
            name="domain"
            placeholder="Domain"
            value={domain}
            onChange={e => this.setState({ domain: e.target.value })}
          />
          {domainError && <span className={severity}>{domainError}</span>}
        </td>
        <td style={{ width: "40%" }}>
          <input
            type="text"
            name="range"
            placeholder="Range"
            value={range}
            onChange={e => this.setState({ range: e.target.value })}
          />
          {rangeError && <span className={severity}>{rangeError}</span>}
        </td>
        <td style={{ width: "20%" }}>
          <button className="info" onClick={this.editTranslationsHandler}>
            Edit
          </button>
          <button
            className="warning"
            onClick={() => removeTranslation(dictionaryUUID, translationUUID)}
          >
            Delete
          </button>
        </td>
      </tr>
    );
  }
}
