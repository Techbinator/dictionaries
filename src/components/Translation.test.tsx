import React from "react";
import Translation from "./Translation";
import { shallow } from "enzyme";
import contextMock from "../helpers/__mocks__/context.mock";

describe("Translation component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(
      <Translation
        translationUUID="58f6b540-4116-11e9-9306-ef359c18301d"
        dictionaryUUID="619ce930-4116-11e9-9306-ef359c18301d"
        translations={contextMock.translations}
        translationsUUID={
          contextMock.dictionaries["619ce930-4116-11e9-9306-ef359c18301d"]
            .translationsUUID
        }
        removeTranslation={contextMock.removeTranslation}
        editTranslation={contextMock.editTranslation}
      />
    );
  });
  it("should render snapshot with default values", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should remove translation", () => {
    wrapper.find(".warning").simulate("click");
    expect(contextMock.removeTranslation).toBeCalledWith(
      "619ce930-4116-11e9-9306-ef359c18301d",
      "58f6b540-4116-11e9-9306-ef359c18301d"
    );
  });

  it("should edit translation", () => {
    wrapper.find('input[name="domain"]').simulate("change", {
      target: { value: "Domain" }
    });
    wrapper.find('input[name="range"]').simulate("change", {
      target: { value: "Range" }
    });
    wrapper.find(".info").simulate("click");
    expect(contextMock.editTranslation).toBeCalledWith({
      uuid: "58f6b540-4116-11e9-9306-ef359c18301d",
      domain: "Domain",
      range: "Range"
    });
  });

  it("should throw validation error", () => {
    const domainInput = wrapper.find('input[name="domain"]');
    const rangeInput = wrapper.find('input[name="range"]');
    domainInput.simulate("change", {
      target: { value: "Domain" }
    });
    wrapper.find('input[name="range"]').simulate("change", {
      target: { value: "Domain" }
    });
    wrapper.find(".info").simulate("click");
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toEqual("Domain must not equal range");
    domainInput.simulate("focus");
    expect(wrapper.state().domainError).toEqual("");
    rangeInput.simulate("focus");
    expect(wrapper.state().rangeError).toEqual("");
  });
});
