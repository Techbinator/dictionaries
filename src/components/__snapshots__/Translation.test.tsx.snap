// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation component should render snapshot with default values 1`] = `
<tr>
  <td
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <input
      name="domain"
      onChange={[Function]}
      placeholder="Domain"
      type="text"
      value="a"
    />
  </td>
  <td
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <input
      name="range"
      onChange={[Function]}
      placeholder="Range"
      type="text"
      value="b"
    />
  </td>
  <td
    style={
      Object {
        "width": "20%",
      }
    }
  >
    <button
      className="info"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="warning"
      onClick={[Function]}
    >
      Delete
    </button>
  </td>
</tr>
`;
